modules/telemetry/telemetry/include/hephaestus/telemetry/struclog.h

```diff
@@ -0,0 +1,187 @@
+};
+
+/// @brief A class that allows easy composition of logs for structured logging.
+///       Example(see also struclog.cpp):
+///       ```cpp
+///         namespace ht=heph::telemetry;
+///         log(Level::INFO, "adding", {"speed", 31.3}, {"tag", "test"});
+///       ```
+///         logs
+///        'level=info hostname=goofy location=struclog.h:123 thread-id=5124 time=2023-12-3T8:52:02+0
+/// module="my module" message="adding" id=12345 speed=31.3 tag="test"'
+///        Remark: We would like to use libfmt with quoted formatting as proposed in
+///                https://github.com/fmtlib/fmt/issues/825 once its included (instead of sstream).
+struct LogEntry {
+  using FieldsT = std::vector<Field<std::string>>;
+  using ClockT = std::chrono::system_clock;
+
+  LogEntry(Level level, MessageWithLocation message);
+
+  /// @brief General loginfo consumer, should be used like LogEntry("my message") << Field{"field", 1234}
+  ///        Converted to string with stringstream.
+  ///
+  /// @tparam T any type that is consumeable by stringstream
+  /// @param key identifier for the logging data
+  /// @param value value of the logging data
+  /// @return LogEntry&&
+
+  template <NonQuotable T>
+  auto operator<<(const Field<T>& field) -> LogEntry&& {
+    std::stringstream ss;
+    ss << field.value;
+    fields.emplace_back(field.key, ss.str());
+
+    return std::move(*this);
+  }
+
+  /// @brief Specialized loginfo consumer for string types so that they are quoted, should be used like
+  ///        LogEntry("my message") << Field{"field", "mystring"}
+  ///
+  /// @tparam T any type that is consumeable by stringstream
+  /// @param key identifier for the logging data
+  /// @param val value of the logging data
+  /// @return LogEntry&&
+
+  template <Quotable S>
+  auto operator<<(const Field<S>& field) -> LogEntry&& {
+    std::stringstream ss;
+    // Pointer decay is wanted here to catch char[n] messages
+    // NOLINTNEXTLINE(cppcoreguidelines-pro-bounds-array-to-pointer-decay)
+    ss << std::quoted(field.value);
+    fields.emplace_back(field.key, ss.str());
+
+    return std::move(*this);
+  }
+
+  Level level;
+  MessageWithLocation msg_with_loc;
+  std::thread::id thread_id;
+  ClockT::time_point time;
+  std::string hostname;
+
+  FieldsT fields;
+};
+
+/// @brief format function from LogEntry to string adheruing to logfmt rules.
+///
+/// @param log
+/// @return std::string
+[[nodiscard]] auto format(const LogEntry& log) -> std::string;
+
+inline auto operator<<(std::ostream& os, const LogEntry& log) -> std::ostream& {
+  os << format(log);
+
+  return os;
+}
+
+/// @brief Interface for sinks that log entries can be sent to.
+struct ILogSink {
+  using Formatter = std::function<std::string(const LogEntry& log)>;
+
+  virtual ~ILogSink() = default;
+
+  /// @brief Main interface to the sink, gets called on log.
+  ///
+  /// @param log_entry
+  virtual void send(const LogEntry& log_entry) = 0;
+};
```

Comment:
please move this into `log_sink.h`

Reply:
done for this and LogEntry (same as metric_sink and metric_record) also renamed struclog.h to log.h